server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java

```diff
@@ -0,0 +1,356 @@
+                "recipients", StringUtils.join(mail.getRecipients()))))
+            .log("PostgresMailRepository stored mail."));
+    }
+
+    private String[] asStringArray(Collection<MailAddress> mailAddresses) {
+        return mailAddresses.stream()
+            .map(MailAddress::asString)
+            .toArray(String[]::new);
+    }
+
+    private Hstore asHstore(Multimap<MailAddress, PerRecipientHeaders.Header> multimap) {
+        return Hstore.hstore(multimap
+            .asMap()
+            .entrySet()
+            .stream()
+            .map(recipientToHeaders -> Pair.of(recipientToHeaders.getKey().asString(),
+                asString(recipientToHeaders.getValue())))
+            .collect(ImmutableMap.toImmutableMap(Pair::getLeft, Pair::getRight)));
+    }
+
+    private String asString(Collection<PerRecipientHeaders.Header> headers) {
+        return StringUtils.join(headers.stream()
+            .map(PerRecipientHeaders.Header::asString)
+            .collect(ImmutableList.toImmutableList()), HEADERS_SEPARATOR);
+    }
+
+    private Hstore asHstore(Stream<Attribute> attributes) {
+        return Hstore.hstore(attributes
+            .flatMap(attribute -> attribute.getValue()
+                .toJson()
+                .map(JsonNode::toString)
+                .map(value -> Pair.of(attribute.getName().asString(), value)).stream())
+            .collect(ImmutableMap.toImmutableMap(Pair::getLeft, Pair::getRight)));
+    }
+
+    public Iterator<MailKey> list(MailRepositoryUrl url) throws MessagingException {
+        return listMailKeys(url)
+            .toStream()
+            .iterator();
+    }
+
+    private Flux<MailKey> listMailKeys(MailRepositoryUrl url) {
+        return postgresExecutor.executeRows(context -> Flux.from(context.select(KEY)
+                .from(TABLE_NAME)
+                .where(URL.eq(url.asString()))))
+            .map(record -> new MailKey(record.get(KEY)));
+    }
+
+    public Mail retrieve(MailKey key, MailRepositoryUrl url) {
+        return postgresExecutor.executeRow(context -> Mono.from(context.select()
+                .from(TABLE_NAME)
+                .where(URL.eq(url.asString()))
+                .and(KEY.eq(key.asString()))))
+            .flatMap(this::toMail)
+            .blockOptional()
+            .orElse(null);
+    }
+
+    private Mono<Mail> toMail(Record record) {
+        return mimeMessageStore.read(toMimeMessagePartsId(record))
+            .map(Throwing.function(mimeMessage -> toMail(record, mimeMessage)));
+    }
+
+    private Mail toMail(Record record, MimeMessage mimeMessage) throws MessagingException {
+        List<MailAddress> recipients = Arrays.stream(record.get(RECIPIENTS))
+            .map(Throwing.function(MailAddress::new))
+            .collect(ImmutableList.toImmutableList());
+
+        PerRecipientHeaders perRecipientHeaders = getPerRecipientHeaders(record);
+
+        List<Attribute> attributes = Hstore.hstore(record.get(ATTRIBUTES, LinkedHashMap.class))
```

Comment:
I don't know why we need call to `Hstore` here
It should not invoked when un-necessary (that created for of Jooq extension)
We can entrySet directly on LinkedHashMap

```java
((LinkedHashMap<String, String>) record.get(PER_RECIPIENT_SPECIFIC_HEADERS, LinkedHashMap.class))
            .entrySet()
```

Reply:
The `ATTRIBUTES` data type is `hstore`. 
`record.get(ATTRIBUTES)` return `Hstore` at compilation time, but return `LinkedHashMap` at run time. Funny, no? (jooq bug I bet)

That is why I casted `LinkedHashMap.class` there.

I see nothing else wrong with that.

Reply:
But I am OK to remove the intermediate Hstore object there.